nep-12.mediawiki

```diff
@@ -0,0 +1,2130 @@
+  toAddress: 'ATaWxfUAiBcQixNPq6TvKHEHPQum9bx79d',
+  asset: '602c79718b16e442de58778e148d0b1084e3b2dffd5de6b7b16cee7969282de7',
+  amount: '0.0001',
+  remark: 'Hash puppy clothing purchase. Invoice#abc123',
+  fee: '0.0001'
+})
+.then(({txid, nodeUrl}: SendOutput) => {
+  console.log('Send transaction success!');
+  console.log('Transaction ID: ' + txid);
+  console.log('RPC node URL: ' + nodeUrl);
+})
+.catch(({type: string, description: string, data: any}) => {
+  switch(type) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case SEND_ERROR:
+      console.log('There was an error when broadcasting this transaction to the network.');
+      break;
+    case MALFORMED_INPUT:
+      console.log('The receiver address provided is not valid.');
+      break;
+    case CANCELED:
+      console.log('The user has canceled this transaction.');
+      break;
+    case INSUFFICIENT_FUNDS:
+      console.log('The user has insufficient funds to execute this transaction.');
+      break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request
+* Validate the inputs
+** From address matches connected account
+** To address is valid
+** Validate asset exists
+** Amount is valid value, and account has enough balance
+* If the "broadcastOverride" input argument is set to True, return the serialized signed transaction to the dApp
+* Else broadcast the RPC request
+* Return the transaction Id and the url of the rpc node which the tx was broadcast to
+
+Example
+<pre>
+{
+  txid: 'ed54fb38dff371be6e3f96e4880405758c07fe6dd1295eb136fe15f311e9ff77',
+  nodeUrl: 'http://seed7.ngd.network:10332',
+}
+</pre>
+
+
+====invoke====
+Execute a contract invocation.
+
+
+'''Method Interface'''
+
+<pre>
+function invoke({
+  scriptHash: string,
+  operation: string,
+  args: Argument[],
+  attachedAssets?: AttachedAssets,
+  fee?: string,
+  assetIntentOverrides?: AssetIntentOverrides,
+  triggerContractVerification?: boolean,
+  network?: string,
+  broadcastOverride?: boolean,
+  txHashAttributes?: TxHashAttribute[],
+}): Promise<InvokeOutput>
+</pre>
+
+
+'''Input arguments'''
+
+<pre>
+interface InvokeArgs {
+  scriptHash: string; // script hash of the smart contract to invoke
+  operation: string; // operation on the smart contract to call
+  args: Argument[]; // any input arguments for the operation
+  fee?: string; // (Optional) The parsed amount of network fee (in GAS) to include with transaction
+  network?: string - Network to submit this request to. If omitted, will default to network the wallet is currently set to.
+  attachedAssets?: AttachedAssets;
+  broadcastOverride?: boolean; // In the case that the dApp would like to be responsible for broadcasting the signed transaction rather than the wallet provider
+
+  assetIntentOverrides?: AssetIntentOverrides;
+  // A hard override of all transaction utxo inputs and outputs.
+  // IMPORTANT: If provided, fee and attachedAssets will be ignored.
```

Comment:
What the meaning of ignoring passed parameters? Better just not attach them, no?

Reply:
The reason is that by providing the UTXOs for `assetIntentOverrides`, the wallet provider assumes that the dapp has accurately calculated the vin and vout values. If the wallet provider were to alter these values, it will disrupt the intention of the dapp.

Reply:
Ok, males sense, dont know why I asked that :joy: